File: app/src/main/java/com/benmohammad/multithreading/demo/designthread/MyBlockingQueue.java
package com.benmohammad.multithreading.demo.designthread;

import java.util.LinkedList;
import java.util.Queue;

class MyBlockingQueue {

    private final Object QUEUE_LOCK = new Object();
    private final int mCapacity;
    private int mCurrentSize = 0;
    private final Queue<Integer> mQueue = new LinkedList<>();

    public MyBlockingQueue(int capacity) {
        this.mCapacity = capacity;
    }

    public void put(int number) {
        synchronized (QUEUE_LOCK) {
            while(mCurrentSize >= mCapacity) {
                try {
                    QUEUE_LOCK.wait();
                } catch (InterruptedException e) {
                    return;
                }
            }
            mQueue.offer(number);
            mCurrentSize++;
            QUEUE_LOCK.notifyAll();
        }
    }

    public int take() {
        synchronized (QUEUE_LOCK) {
            while(mCurrentSize <= 0) {
                try {
                    QUEUE_LOCK.wait();
                } catch (InterruptedException e) {
                    return 0;
                }
            }

            mCurrentSize--;
            QUEUE_LOCK.notifyAll();
            return mQueue.poll();

        }
    }

}

File: app/src/main/java/com/benmohammad/multithreading/demo/designcoroutines/MyBlockingQueue.kt
package com.benmohammad.multithreading.demo.designcoroutines

import java.util.*
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

internal class MyBlockingQueue(private val capacity : Int) {

    private val reentrantLock = ReentrantLock()
    private val lockCondition = reentrantLock.newCondition()
    private val queue = LinkedList<Int>()
    private var currentSize = 0

    fun put(number : Int) {
        reentrantLock.withLock {
            while(currentSize >= capacity) {
                try {
                    lockCondition.await()
                } catch (e: InterruptedException) {
                    return
                }
            }
            queue.offer(number)
            currentSize++
            lockCondition.signalAll()
        }
    }

    fun take(): Int {
        reentrantLock.withLock {
            while (currentSize <= 0) {
                try {
                    lockCondition.await()
                } catch (e: InterruptedException) {
                    return 0
                }
            }
            currentSize--
            lockCondition.signalAll()
            return queue.poll()
        }
    }


}